### PR TITLE
[AIDAPP-302]: Prevent users from being able to edit service requests of types if they are not managers

### PR DESCRIPTION
**Pull request by ankit-canyon**
### Ticket(s) or GitHub Issue

- https://canyongbs.atlassian.net/browse/AIDAPP-302

### Technical Description

> Prevent users from being able to create/edit/delete service requests of types if they are not managers.

### Any deployment steps required?

> No.

### Are any Feature Flags and/or Data Migrations that can eventually be removed Added?

> No.

_______________________________________________

### CODE REVIEW

File: app-modules/contact/src/Filament/Resources/ContactResource/RelationManagers/ServiceRequestsRelationManager.php [hunk 1]
```diff
@@ -69,6 +69,17 @@ public function table(Table $table): Table
...
             ->recordTitleAttribute('id')
+            ->modifyQueryUsing(function ($query) {
+                $query->when(! auth()->user()->hasRole('authorization.super_admin'), function (Builder $q) {
+                    return $q->whereHas('priority.type.managers', function (Builder $query): void {
+                        $query->where('teams.id', auth()->user()->teams()->first()?->getKey());
+                    })->orWhereHas('priority.type.auditors', function (Builder $query): void {
+                        $query->where('teams.id', auth()->user()->teams()->first()?->getKey());
+                    })->whereHas('respondent', function (Builder $query) {
+                        $query->where('respondent_id', $this->getOwnerRecord()->getKey());
+                    });
```
Comment (Orrison) [inline app-modules/contact/src/Filament/Resources/ContactResource/RelationManagers/ServiceRequestsRelationManager.php:80]: Just curious, is this `whereHas` needed? I would assume the relation manager query already has this scoping. Or does apply the above whereHas overwrite that, making this needed?

File: app-modules/service-management/src/Policies/ServiceRequestPolicy.php [hunk 2]
```diff
@@ -96,6 +96,14 @@ public function view(Authenticatable $authenticatable, ServiceRequest $serviceRe
 
     public function create(Authenticatable $authenticatable): Response
     {
+        if (! auth()->user()->hasRole('authorization.super_admin')) {
+            $team = auth()->user()->teams()->first();
+
+            if (! $team?->managableServiceRequestTypes()->exists() && ! $team?->auditableServiceRequestTypes()->exists()) {
+                return Response::deny("You don't have permission to view this service request because you're not an auditor or manager.");
```
Comment (Orrison) [inline app-modules/service-management/src/Policies/ServiceRequestPolicy.php:103]:
```suggestion
                return Response::deny("You don't have permission to create service requests because you're not an auditor or manager of any service request types.");
```

File: app-modules/service-management/src/Policies/ServiceRequestPolicy.php [hunk 3]
```diff
@@ -112,6 +120,18 @@ public function update(Authenticatable $authenticatable, ServiceRequest $service
...
+        if (! auth()->user()->hasRole('authorization.super_admin')) {
+            $team = auth()->user()->teams()->first();
+
+            if (! $serviceRequest?->priority?->type?->managers()->exists() && ! $serviceRequest?->priority?->type?->auditors()->exists()) {
+                return Response::deny("You don't have permission to update this service request because you're not an auditor or manager.");
+            }
+
+            if (! $serviceRequest?->priority?->type?->managers->contains('id', $team?->getKey()) && ! $serviceRequest?->priority?->type?->auditors->contains('id', $team?->getKey())) {
+                return Response::deny("You don't have permission to update this service request because you're not an auditor or manager.");
+            }
```
Comment (Orrison) [inline app-modules/service-management/src/Policies/ServiceRequestPolicy.php:132]: I think this can all be combined into the following and reduce the number of queries:

```suggestion
            if (! $serviceRequest?->priority?->type?->managers?->contains('id', $team?->getKey())) {
                return Response::deny("You don't have permission to update this service request because you're not a manager of it's type.");
            }
```

Also, I have removed the auditor check as per the ticket requirements, you should only be able to edit it if you are a manager.

File: app-modules/service-management/src/Policies/ServiceRequestPolicy.php [hunk 4]
```diff
@@ -124,6 +144,18 @@ public function delete(Authenticatable $authenticatable, ServiceRequest $service
...
+        if (! auth()->user()->hasRole('authorization.super_admin')) {
+            $team = auth()->user()->teams()->first();
+
+            if (! $serviceRequest?->priority?->type?->managers()->exists() && ! $serviceRequest?->priority?->type?->auditors()->exists()) {
+                return Response::deny("You don't have permission to delete this service request because you're not an auditor or manager.");
+            }
+
+            if (! $serviceRequest?->priority?->type?->managers->contains('id', $team?->getKey()) && ! $serviceRequest?->priority?->type?->auditors->contains('id', $team?->getKey())) {
+                return Response::deny("You don't have permission to delete this service request because you're not an auditor or manager.");
+            }
```
Comment (Orrison) [inline app-modules/service-management/src/Policies/ServiceRequestPolicy.php:156]: The same feedback as the update method above.

File: app-modules/service-management/src/Policies/ServiceRequestPolicy.php [hunk 5]
```diff
@@ -136,6 +168,18 @@ public function restore(Authenticatable $authenticatable, ServiceRequest $servic
...
+        if (! auth()->user()->hasRole('authorization.super_admin')) {
+            $team = auth()->user()->teams()->first();
+
+            if (! $serviceRequest?->priority?->type?->managers()->exists() && ! $serviceRequest?->priority?->type?->auditors()->exists()) {
+                return Response::deny("You don't have permission to restore this service request because you're not an auditor or manager.");
+            }
+
+            if (! $serviceRequest?->priority?->type?->managers->contains('id', $team?->getKey()) && ! $serviceRequest?->priority?->type?->auditors->contains('id', $team?->getKey())) {
+                return Response::deny("You don't have permission to restore this service request because you're not an auditor or manager.");
+            }
```
Comment (Orrison) [inline app-modules/service-management/src/Policies/ServiceRequestPolicy.php:180]: The same feedback as the update method above.

File: app-modules/service-management/src/Policies/ServiceRequestPolicy.php [hunk 6]
```diff
@@ -148,6 +192,18 @@ public function forceDelete(Authenticatable $authenticatable, ServiceRequest $se
...
+        if (! auth()->user()->hasRole('authorization.super_admin')) {
+            $team = auth()->user()->teams()->first();
+
+            if (! $serviceRequest?->priority?->type?->managers()->exists() && ! $serviceRequest?->priority?->type?->auditors()->exists()) {
+                return Response::deny("You don't have permission to permanently delete this service request because you're not an auditor or manager.");
+            }
+
+            if (! $serviceRequest?->priority?->type?->managers->contains('id', $team?->getKey()) && ! $serviceRequest?->priority?->type?->auditors->contains('id', $team?->getKey())) {
+                return Response::deny("You don't have permission to permanently delete service request because you're not an auditor or manager.");
+            }
```
Comment (Orrison) [inline app-modules/service-management/src/Policies/ServiceRequestPolicy.php:204]: The same feedback as the update method above.

File: app-modules/service-management/src/Policies/ServiceRequestPolicy.php [hunk 7]
```diff
@@ -96,6 +96,14 @@ public function view(Authenticatable $authenticatable, ServiceRequest $serviceRe
 
     public function create(Authenticatable $authenticatable): Response
     {
+        if (! auth()->user()->hasRole('authorization.super_admin')) {
+            $team = auth()->user()->teams()->first();
+
+            if (! $team?->managableServiceRequestTypes()->exists() && ! $team?->auditableServiceRequestTypes()->exists()) {
```
Comment (Orrison) [inline app-modules/service-management/src/Policies/ServiceRequestPolicy.php:102]: Let's remove the auditor check here. Only managers should be able to create.

File: app-modules/service-management/src/Filament/Resources/ServiceRequestResource/Pages/CreateServiceRequest.php [hunk 8]
```diff
@@ -83,7 +84,15 @@ public function form(Form $form): Form
                 Grid::make()
                     ->schema([
                         Select::make('type_id')
-                            ->options(ServiceRequestType::pluck('name', 'id'))
+                            ->options(ServiceRequestType::when(! auth()->user()->hasRole('authorization.super_admin'), function (Builder $query) {
+                                $query->whereHas('managers', function (Builder $query): void {
+                                    $query->where('teams.id', auth()->user()->teams()->first()?->getKey());
+                                })->orWhereHas('auditors', function (Builder $query): void {
+                                    $query->where('teams.id', auth()->user()->teams()->first()?->getKey());
+                                });
```
Comment (Orrison) [inline app-modules/service-management/src/Filament/Resources/ServiceRequestResource/Pages/CreateServiceRequest.php:92]: Let's remove the auditor check here. You should only be able to create ones you are a manager of.

File: app-modules/service-management/src/Filament/Resources/ServiceRequestResource/Pages/EditServiceRequest.php [hunk 9]
```diff
@@ -99,12 +100,20 @@ public function form(Form $form): Form
                                         fn (ServiceRequest $record) => ServiceRequestType::withTrashed()
                                             ->whereKey($record->priority?->type_id)
                                             ->orWhereNull('deleted_at')
+                                            ->when(! auth()->user()->hasRole('authorization.super_admin'), function (Builder $query) {
+                                                $query->whereHas('managers', function (Builder $query): void {
+                                                    $query->where('teams.id', auth()->user()->teams()->first()?->getKey());
+                                                })->orWhereHas('auditors', function (Builder $query): void {
+                                                    $query->where('teams.id', auth()->user()->teams()->first()?->getKey());
+                                                });
```
Comment (Orrison) [inline app-modules/service-management/src/Filament/Resources/ServiceRequestResource/Pages/EditServiceRequest.php:108]: Let's remove the auditor check here. You should only be able to create ones you are a manager of.

File: app-modules/service-management/src/Rules/ManagedServiceRequestType.php [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+        $team = auth()->user()->teams()->first();
+
+        $isManager = ServiceRequestType::where('id', $value)
+            ->whereHas('managers', function ($query) use ($team) {
+                $query->where('teams.id', $team?->getKey());
+            })
+            ->orWhereHas('auditors', function ($query) use ($team) {
+                $query->where('teams.id', $team?->getKey());
+            })
```
Comment (Orrison) [inline app-modules/service-management/src/Rules/ManagedServiceRequestType.php:64]: Let's remove the auditor check here. You should only be able to create ones you are a manager of.

File: app-modules/service-management/tests/ServiceRequest/CreateServiceRequestTest.php [hunk 11]
```diff
@@ -239,3 +273,167 @@
 
     expect($serviceRequest->division->id)->toEqual($request['division_id']);
 });
+
+test('cannot create service requests if user is not an auditor or manager of any service request type', function () {
```
Comment (Orrison) [inline app-modules/service-management/tests/ServiceRequest/CreateServiceRequestTest.php:277]: This should test that only managers can create

File: app-modules/service-management/tests/ServiceRequest/CreateServiceRequestTest.php [hunk 12]
```diff
@@ -239,3 +273,167 @@
...
+    livewire(CreateServiceRequest::class)
+        ->assertFormFieldExists('type_id', function (Select $field) use ($serviceRequestTypesWithManagers, $serviceRequestTypesWithoutManagers): bool {
+            $options = $field->getOptions();
+
+            return in_array($serviceRequestTypesWithManagers->getKey(), array_keys($options)) &&
+                   ! in_array($serviceRequestTypesWithoutManagers->getKey(), array_keys($options));
+        });
+});
+
+test('create service requests if user is an auditor or manager of any service request type', function () {
```
Comment (Orrison) [inline app-modules/service-management/tests/ServiceRequest/CreateServiceRequestTest.php:348]: The name and/or content should be changed to only focus on managers. Only managers can create

File: app-modules/service-management/tests/ServiceRequest/CreateServiceRequestTest.php [hunk 13]
```diff
@@ -239,3 +273,167 @@
...
+    assertCount(1, ServiceRequest::all());
+
+    assertDatabaseHas(ServiceRequest::class, $request->except(['division_id', 'respondent_id', 'type_id'])->toArray());
+
+    $serviceRequest = ServiceRequest::first();
+
+    expect($serviceRequest->division->id)->toEqual($request['division_id']);
+});
+
+test('validate service requests type if user is not an auditor or manager of any service request type', function () {
```
Comment (Orrison) [inline app-modules/service-management/tests/ServiceRequest/CreateServiceRequestTest.php:397]: Only managers can create, ensure that is true in the name of this test and content of this test if present

File: app-modules/service-management/tests/ServiceRequest/EditServiceRequestTest.php
Comment (Orrison) [inline app-modules/service-management/tests/ServiceRequest/EditServiceRequestTest.php:1]: Ensure the tests in this file cover that, only managers can edit/create, not auditors.

File: app-modules/service-management/tests/ServiceRequest/ViewServiceRequestTest.php [hunk 14]
```diff
@@ -167,8 +167,6 @@
     $user->givePermissionTo('service_request.view-any');
     $user->givePermissionTo('service_request.*.view');
 
-    $user->refresh();
-
```
Comment (Orrison) [inline app-modules/service-management/tests/ServiceRequest/ViewServiceRequestTest.php:171]: Why was this removed?

**Review by Orrison — CHANGES_REQUESTED** [review]

File: app-modules/service-management/tests/ServiceRequest/ViewServiceRequestTest.php (hunk 14 above)
Comment (ankit-canyon) [reply app-modules/service-management/tests/ServiceRequest/ViewServiceRequestTest.php:171]: Added again !

File: app-modules/contact/src/Filament/Resources/ContactResource/RelationManagers/ServiceRequestsRelationManager.php (hunk 1 above)
Comment (ankit-canyon) [reply app-modules/contact/src/Filament/Resources/ContactResource/RelationManagers/ServiceRequestsRelationManager.php:80]: Yep ! It's actually overriding the method. Services requests which not related to specific contact are also showing in the list. Hence by applying custom query I prevented that issue.

**Review by Orrison — APPROVED** [review]
